test/integration/controllermanager/serving/healthcheck_test.go

```diff
@@ -0,0 +1,236 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serving
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"testing"
+
+	"k8s.io/klog/v2/ktesting"
+	kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	kubecontrollermanageresting "k8s.io/kubernetes/cmd/kube-controller-manager/app/testing"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+func TestHealthEndpoints(t *testing.T) {
+	server, configStr, _, err := startTestAPIServer(t)
+	if err != nil {
+		t.Fatalf("Failed to start kube-apiserver server: %v", err)
+	}
+	defer server.TearDownFn()
+
+	apiserverConfig, err := os.CreateTemp("", "kubeconfig")
+	if err != nil {
+		t.Fatalf("Failed to create config file: %v", err)
+	}
+	defer func() {
+		_ = os.Remove(apiserverConfig.Name())
+	}()
+	if _, err = apiserverConfig.WriteString(configStr); err != nil {
+		t.Fatalf("Failed to write kubeconfig file: %v", err)
+	}
+
+	brokenConfigStr := strings.ReplaceAll(configStr, "127.0.0.1", "127.0.0.2")
+	brokenConfig, err := os.CreateTemp("", "kubeconfig")
+	if err != nil {
+		t.Fatalf("Failed to create kubeconfig file: %v", err)
+	}
+	if _, err := brokenConfig.WriteString(brokenConfigStr); err != nil {
+		t.Fatalf("Failed to write kubeconfig file: %v", err)
+	}
+	defer func() {
+		_ = os.Remove(brokenConfig.Name())
+	}()
+
+	tests := []struct {
+		name             string
+		path             string
+		useBrokenConfig  bool
+		wantResponseCode int
+	}{
+		{
+			"/healthz",
+			"/healthz",
+			false,
+			http.StatusOK,
+		},
+		{
+			"/livez",
+			"/livez",
+			false,
+			http.StatusOK,
+		},
+		{
+			"/livez with ping check",
+			"/livez/ping",
+			false,
+			http.StatusOK,
+		},
+		{
+			"/readyz",
+			"/readyz",
+			false,
+			http.StatusOK,
+		},
+		{
+			"/readyz with sched-handler-sync",
+			"/readyz/kcm-handler-sync",
+			false,
+			http.StatusOK,
+		},
+		{
+			"/readyz with shutdown",
+			"/readyz/shutdown",
+			false,
+			http.StatusOK,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt := tt
+			_, ctx := ktesting.NewTestContext(t)
+
+			configFile := apiserverConfig.Name()
+			if tt.useBrokenConfig {
+				configFile = brokenConfig.Name()
+			}
+			result, err := kubecontrollermanageresting.StartTestServer(
+				t,
+				ctx,
+				[]string{"--controllers=", "--kubeconfig", configFile, "--leader-elect=false", "--authorization-always-allow-paths", tt.path},
+			)
+
+			if err != nil {
+				t.Fatalf("Failed to start kube-controller-manager server: %v", err)
+			}
+			if result.TearDownFn != nil {
+				defer result.TearDownFn()
+			}
+
+			client, base, err := clientAndURLFromTestServer(result)
+			if err != nil {
+				t.Fatalf("Failed to get client from test server: %v", err)
+			}
+			req, err := http.NewRequest(http.MethodGet, base+tt.path, nil)
+			if err != nil {
+				t.Fatalf("failed to request: %v", err)
```

Comment:
use consistent uppercase in t.Failed / t.Error e.g. `failed ..` -> `Failed ...`
more of that below.